### PR TITLE
zenfs: better collaborative GC & better metrics

### PR DESCRIPTION
**Pull request by royguo**
This MR contains **FOUR** major changes, most of the changes are not related to the ext4 code path:

---

**Commit 1: enhancement ZNS GC**
_Authored by @levisonchen and @yapple, directed by @mm304321141_ 
- Add weighted hints for TerarkDB to pick compaction files smarter
- TerarkDB could get zenfs's snapshot to understand which files are placed in the zone with the most garbage
---
**Commit 2: Add Support for Official ZenFS**
_Authored by liuruicheng.222@bytedance.com, directed by @royguo_ 
- Switch zenfs from `bytedance/bzbd` to `wdc/zenfs`
- Add decoupled metrics tracking metrics
---
**Commit 3: fixed a bug in GetStat**
- Minor changes
---
**Commit 4: zenfs: optimize collaborative GC & metrics API**
_Authored by @royguo_ 
- Update ZenFS to the latest version
- Refine metrics framework
- Add better collaborative GC implement (with zenfs extent migration support)

### CODE REVIEW

File: db/db_impl.cc [hunk 1]
```diff
@@ -1086,14 +1120,25 @@ void DBImpl::ScheduleZNSGC() {
 
   chash_set<uint64_t> mark_for_gc;
 
-  if (initial_db_options_.zenfs_gc_ratio <= 0.0 ||
-      initial_db_options_.zenfs_gc_ratio >= 1.0) {
-    // GC is not enabled
+  double low_r = initial_db_options_.zenfs_low_gc_ratio;
+  double high_r = initial_db_options_.zenfs_high_gc_ratio;
+  double force_r = initial_db_options_.zenfs_force_gc_ratio;
```
Comment (mm304321141) [inline db/db_impl.cc]: 这三个参数，从 `mutable_db_options_` 里面获取。（注意需要在锁内读取）

File: db/db_impl.cc [hunk 2]
```diff
@@ -1086,14 +1120,25 @@ void DBImpl::ScheduleZNSGC() {
...
+
+  if (!(0.0 < low_r && low_r <= high_r && high_r <= force_r &&
```
Comment (mm304321141) [inline db/db_impl.cc]: 这三个参数，在 `DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src)` 中检验，修整。
在 `db_impl_open.cc` 文件

**Comment by FuturisticWater** [comment]
Could you please update the MR description with a brief summary of the objective, high-level design and implementation strategy? This would greatly help readers understand and evaluate this MR. Thanks!

**Comment by FuturisticWater** [comment]
> Could you please update the MR description with a brief summary of the objective, high-level design and implementation strategy? This would greatly help readers understand and evaluate this MR. Thanks!

I just saw that each individual commit actually has a nice description, so maybe just summarize the list of changes here in the MR description.

Also, is it possible to merge the commits one-by-one?

**Comment by royguo** [comment]
Hi, @hw77-git , Sorry for the inconvenience.

These four PRs are a historical problem: the 2nd & 4th commit overrides some code of the first one, so reviewing them one by one will make reviewers confused.

I believe this is the last MR that contains multiple changes, we will follow the rules in the next PRs.

File: db/db_impl.cc (hunk 2 above)
Comment (royguo) [reply db/db_impl.cc]: fixed, PTAL

File: db/db_impl.cc (hunk 1 above)
Comment (royguo) [reply db/db_impl.cc]: Fixed, PTAL

**Comment by mm304321141** [comment]
LGTM

**Pull request merged**